core-bundle/src/EventListener/DataContainer/ContentCompositionListener.php

```diff
@@ -0,0 +1,268 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\Backend;
+use Contao\BackendUser;
+use Contao\CoreBundle\Framework\Adapter;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Routing\Content\PageProviderInterface;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\DataContainer;
+use Contao\Image;
+use Contao\LayoutModel;
+use Contao\PageModel;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use Terminal42\ServiceAnnotationBundle\ServiceAnnotationInterface;
+
+class ContentCompositionListener implements ServiceAnnotationInterface
```

Comment:
this is because the listener does not only handle `tl_page`, it also handles `tl_article`. It handles everything related to "content composition" (as described in https://gadgetopia.com/post/9429/). I don't think we should have one listener per data container but rather have listeners by _functionality_.

Reply:
Why do we need to refactor the code at all? Why not keep it where it is now?

Reply:
We had some related discussion about this in #1762 as well. It would be helpful to decide on a definite structure for DCA callbacks within the core :)